### ISSUE TITLE
[SF Migration] Provide a rollback hook for FormHandler

### ISSUE THREAD

**Issue by matks** [PrestaShop/PrestaShop#14193]
Type: tech-issue

https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Form/IdentifiableObject/Handler/FormHandler.php#L168
```
$this->hookDispatcher->dispatchWithParameters(
            'actionBeforeCreate' . Container::camelize($form->getName()) . 'FormHandler', [
                'form_data' => &$data,
            ]
        );
        $id = $this->dataHandler->create($data);
        $this->hookDispatcher->dispatchWithParameters('actionAfterCreate' . Container::camelize($form->getName()) . 'FormHandler', [
            'id' => $id,
            'form_data' => &$data,
        ]);
```

We have 2 hooks there:
- `actionBeforeCreate{formName}FormHandler` allows module developer to perform an action before the `$dataHandler` creates the data
- `actionAfterCreate{formName}FormHandler` allows module developer to perform an action after the `$dataHandler` creates the data and can use the ID of the data created by the `$dataHandler`

What happens if something goes bad ?

If something bad happens when `actionBeforeCreate{formName}FormHandler` attempts to persist data (for example), it's fine, it should throw an Exception and nothing has been created.

But what if something bad happens when `$this->dataHandler->create($data);` is called ? It might mean the data is not created but the hook alreadid did its job. So database might now be in an irrelevant state where half the data has been created.

If the module is supposed to add data in another SQL table `ps_customers_extra` that should always be synchronized with `ps_customers` then we have the new module data in  `ps_customers_extra` but `ps_customers` has not the new data. The tables are not synchronized anymore.

As suggested by @eternoendless, we should provide module developers one way to handle this usecase to "rollback" their modification, with a dedicated hook:

```
        $this->hookDispatcher->dispatchWithParameters(
            'actionBeforeCreate' . Container::camelize($form->getName()) . 'FormHandler', [
                'form_data' => &$data,
            ]
        );
        try {
            $id = $this->dataHandler->create($data);
        } catch (\Exception $e) {
            $this->hookDispatcher->dispatchWithParameters(
            'actionAfterCreateError' . Container::camelize($form->getName()) . 'FormHandler', [
                'form_data' => &$data,
                'exception' => $e
            ]);

            throw $e;
        }
        $this->hookDispatcher->dispatchWithParameters('actionAfterCreate' . Container::camelize($form->getName()) . 'FormHandler', [
            'id' => $id,
            'form_data' => &$data,
        ]);
```

**Comment by matks** [comment]
@v4lux you might be interested in this strategy being applied everywhere ?

**Comment by Nakahiru** [comment]
Indeed! :) It would be nice to have this behavior as mush possible on every hook. For example, the hook actionOrderSlipAdd -> This hook is executed when the merchant process to a refund on an order. Payment modules may eventually call a payment API to also make the refund on the platform payment side. If the API returns an error it could be nice to rollback all actions made by the hook.
